src/apify/_actor.py

```diff
@@ -839,12 +841,29 @@ async def reboot(
             self.log.error('Actor.reboot() is only supported when running on the Apify platform.')
             return
 
+        if self._is_rebooting:
+            self.log.debug('Actor is already rebooting, skipping the additional reboot call.')
+            return
+
+        self._is_rebooting = True
+
         if not custom_after_sleep:
             custom_after_sleep = self._configuration.metamorph_after_sleep
 
-        self._event_manager.emit(event=Event.PERSIST_STATE, event_data=EventPersistStateData(is_migrating=True))
+        # Call all the listeners for the PERSIST_STATE and MIGRATING events, and wait for them to finish.
+        # We can't just emit the events and wait for all listeners to finish,
+        # because this method might be called from an event listener itself, and we would deadlock.
+        persist_state_listeners = chain.from_iterable(
+            (self._event_manager._listeners_to_wrappers[Event.PERSIST_STATE] or {}).values()  # noqa: SLF001
+        )
+        migrating_listeners = chain.from_iterable(
+            (self._event_manager._listeners_to_wrappers[Event.MIGRATING] or {}).values()  # noqa: SLF001
+        )
 
-        await self._event_manager.__aexit__(None, None, None)
+        await asyncio.gather(
+            *[listener(EventPersistStateData(is_migrating=True)) for listener in persist_state_listeners],
+            *[listener(EventMigratingData()) for listener in migrating_listeners],
```

Comment:
I might be bringing up something we've already discussed, but what if we only emitted `EventPersistData(is_migrating=True)` and skipped the "migrating" event? If it's just for parity with the JS SDK, maybe we could just... not port it here?

Reply:
I think @B4nan said that in JS crawlee, crawlers are listening for the `migrating` event and pausing once they receive it. I suppose this behavior should be consistent between JS and Python.

Reply:
Right, it makes sense to make sure that those get called as well. Could you drop a comment about that in the code?

Reply:
Sure, will do